### PR TITLE
implemented simplest markdown documentation writer

### PR DESCRIPTION
**Pull request by OleksandrKucherenko**
Output Preview

Preview of Output

### CODE REVIEW

**Comment by OleksandrKucherenko** [comment]
@felipecsl ^

**Comment by OleksandrKucherenko** [comment]
needed review...

File: deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/Documentor.java [hunk 1]
```diff
@@ -122,4 +135,92 @@ private static String formatJavaDoc(String str) {
...
+      }
+
+      writer.flush();
+    }
+  }
+
+  /**
+   * Old documentation format.
+   */
+  static class GenericWriter implements DocumetationWriter {
```
Comment (felipecsl) [inline deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/Documentor.java]: can you move this to its own file please?

File: deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/Documentor.java [hunk 2]
```diff
@@ -122,4 +135,92 @@ private static String formatJavaDoc(String str) {
...
+    @Override
+    public void write(@Nonnull final ProcessingEnvironment env,
+                      @Nonnull final PrintWriter writer,
+                      @Nonnull final List<DeepLinkAnnotatedElement> elements) {
```
Comment (felipecsl) [inline deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/Documentor.java]: Please omit `@Nonnull` annotations since they are unnecessary

File: deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/Documentor.java [hunk 3]
```diff
@@ -122,4 +135,92 @@ private static String formatJavaDoc(String str) {
...
+  /**
+   * GitHub markdown format.
+   *
+   * @see <a href="https://guides.github.com/features/mastering-markdown/">Mastering Markdown</a>
+   * @see <a href="https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet">Markdown
+   * Cheatsheet</a>
+   * @see <a href="https://github.com/JFormDesigner/markdown-writer-fx">Markdown Writer FX</a>
+   * @see <a href="https://stackedit.io/">Stack Edit</a>
+   */
+  static class MarkdownWriter implements DocumetationWriter {
```
Comment (felipecsl) [inline deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/Documentor.java]: move to its own file

File: deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/Documentor.java [hunk 4]
```diff
@@ -122,4 +135,92 @@ private static String formatJavaDoc(String str) {
...
+   * Implement this interface if you want to provide own documentation writer.
+   */
+  interface DocumetationWriter {
+    /**
+     * Compose documentation with help of provided environment, writer and collection of
+     * found deeplink elements.
+     */
+    void write(@Nonnull final ProcessingEnvironment env,
+               @Nonnull final PrintWriter writer,
+               @Nonnull final List<DeepLinkAnnotatedElement> elements);
```
Comment (felipecsl) [inline deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/Documentor.java]: remove non-null annotations

File: deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/Documentor.java [hunk 5]
```diff
@@ -30,20 +36,32 @@
...
+   */
+  private static final Map<String, DocumetationWriter> EXTENSIONS =
+      new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
 
   private final ProcessingEnvironment processingEnv;
   private final Messager messager;
 
-  @Nullable private File file;
+  @Nullable
+  private File file;
```
Comment (felipecsl) [inline deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/Documentor.java]: undo this change

File: deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/Documentor.java [hunk 6]
```diff
@@ -30,20 +36,32 @@
...
   private static final String PARAM = "@param";
   private static final String RETURN = "@return";
-  @VisibleForTesting static final String DOC_OUTPUT_PROPERTY_NAME = "deepLinkDoc.output";
+  @VisibleForTesting
+  static final String DOC_OUTPUT_PROPERTY_NAME = "deepLinkDoc.output";
+  /**
+   * Known extensions.
+   */
+  private static final Map<String, DocumetationWriter> EXTENSIONS =
+      new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
```
Comment (felipecsl) [inline deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/Documentor.java]: lets keep it simple - no need to maintain a map of extensions - just hardcode it below

File: deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/Documentor.java [hunk 7]
```diff
@@ -122,4 +135,92 @@ private static String formatJavaDoc(String str) {
...
+      final String format = "| %1$s | %2$s | %3$s | %4$s |";
+
+      // publish lines
+      for (DeepLinkAnnotatedElement element : elements) {
+        final String uri = element.getUri();
+        final String doc = formatJavaDoc(utils.getDocComment(element.getElement()));
+        final String embeddedComments = null == doc ? "" : doc;
+        final boolean isMethod = element.getAnnotationType().equals(DeepLinkEntry.Type.METHOD);
+        final String methodName = isMethod ? element.getMethod() : "";
+        final Name simpleName = element.getAnnotatedElement().getSimpleName();
```
Comment (felipecsl) [inline deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/Documentor.java]: no need to make these variables final

File: README.md [hunk 8]
```diff
@@ -270,6 +270,22 @@ The documentation will be generated in the following format:
...
+ tasks.withType(JavaCompile) {
+   options.compilerArgs << "-AdeepLinkDoc.output=${buildDir}/doc/deeplinks.md"
+ }
+ ```
+
+`.md` extension triggers documentation creation in GitHub markdown syntax.
+
+![Preview of Output](http://i.imgur.com/V4qEGW0.png)
+
+
```
Comment (felipecsl) [inline README.md]: undo readme changes since we can only update the readme once the stable release with this functionality has gone out

File: deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/Documentor.java (hunk 2 above)
Comment (OleksandrKucherenko) [reply deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/Documentor.java]: need here a little details: why its unnecessary?! I always think it helps IDE to detect wrong parameters. `NonNull, Nullable, StringRes, etc.`

File: deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/Documentor.java (hunk 7 above)
Comment (OleksandrKucherenko) [reply deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/Documentor.java]: final == readonly, immutable ... also its a hint to JIT/compiler to inline code for faster execution... I can be wrong of course, but I think that is one from many best practices of Java developers.
